pkg/codegen/dotnet/gen_program_test.go

```diff
@@ -1,11 +1,97 @@
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
+
 	"github.com/pulumi/pulumi/pkg/v3/codegen/internal/test"
+	"github.com/pulumi/pulumi/pkg/v3/testing/integration"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/executable"
 )
 
 func TestGenerateProgram(t *testing.T) {
-	test.TestProgramCodegen(t, "dotnet", GenerateProgram)
+	test.TestProgramCodegen(t,
+		test.ProgramLangConfig{
+			Language:   "dotnet",
+			Extension:  "cs",
+			OutputFile: "MyStack.cs",
+			Check:      checkDotnet,
+			GenProgram: GenerateProgram,
+		},
+	)
+}
+
+func checkDotnet(t *testing.T, path string) {
+	var err error
+	dir := filepath.Dir(path)
+
+	ex, err := executable.FindExecutable("dotnet")
+	assert.NoError(t, err, "Failed to find dotnet executable")
+
+	projectFile := filepath.Join(dir, filepath.Base(dir)+".csproj")
+	if _, err := ioutil.ReadFile(projectFile); os.IsNotExist(err) {
+		defer func() {
+			err = os.Remove(projectFile)
+			assert.NoError(t, err, "Failed to delete project file")
+			err = os.Remove(filepath.Join(dir, "Program.cs"))
+			assert.NoError(t, err, "Failed to delete C# project main")
+		}()
+		err = integration.RunCommand(t, "create dotnet project",
+			[]string{ex, "new", "console"}, dir, &integration.ProgramTestOptions{})
+		assert.NoError(t, err, "Failed to create C# project")
+	}
+
+	// Add dependencies (based on directory name)
+	if pkg, pkgVersion := packagesFromTestName(filepath.Base(dir)); pkg != "" {
+		err = integration.RunCommand(t, "create dotnet project",
+			[]string{ex, "add", "package", pkg, "--version", pkgVersion},
+			dir, &integration.ProgramTestOptions{})
+		assert.NoError(t, err, "Failed to add dependency %q %q", pkg, pkgVersion)
+	} else {
+		err = integration.RunCommand(t, "add sdk ref",
+			[]string{ex, "add", "reference", "../../../../../../sdk/dotnet/Pulumi"},
+			dir, &integration.ProgramTestOptions{})
+		assert.NoError(t, err, "Failed to dotnet sdk package reference")
+	}
+
+	// Clean up build result
+	defer func() {
+		err = os.RemoveAll(filepath.Join(dir, "bin"))
+		assert.NoError(t, err, "Failed to remove bin result")
+		err = os.RemoveAll(filepath.Join(dir, "obj"))
+		assert.NoError(t, err, "Failed to remove obj result")
+	}()
+	err = integration.RunCommand(t, "dotnet build",
+		[]string{ex, "build", "--nologo"}, dir, &integration.ProgramTestOptions{})
+	assert.NoError(t, err, "Failed to build dotnet project")
+}
+
+// packagesFromName attempts to figure out what package should be imported from
+// the name of a test.
+//
+// Example:
+// 	"aws-eks-pp" => ("Pulumi.Aws", 4.21.1)
+// 	"azure-sa-pp" => ("Pulumi.Azure", 4.21.1)
+// 	"resource-options-pp" => ("","")
+//
+// Note: While we could instead do this by using the generateMetaData function
+// for each language, we are trying not to expand the functionality under test.
+func packagesFromTestName(name string) (string, string) {
+	if strings.Contains(name, "aws") {
+		return "Pulumi.Aws", "4.21.1"
```

Comment:
Hmm, this is interesting.. Do these tests truly depend on these providers.

Reply:
They depend on them for type information, so they really do depend on them. It's unfortunate but I can't think of a way around it right now without rewriting the tests.

Reply:
I'd really prefer that we not do this via string matching on filenames:
- it imposes implicit structure
- it's fragile with respect to the contents of the tests

Instead, I'd rather we drive this from the program code generators themselves. However, I realize that's a bigger change, so I think that this is okay provided that we file an issue to come back and clean this up.